File: gtsdbDriver.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strconv"
	"strings"
	"time"
)

// TSDBClient struct remains the same as in the previous example
type TSDBClient struct {
	conn net.Conn
}

// NewTSDBClient creates a new TSDB client
func NewTSDBClient(address string) (*TSDBClient, error) {
	conn, err := net.Dial("tcp", address)
	if err != nil {
		return nil, err
	}
	return &TSDBClient{conn: conn}, nil
}

// Close closes the connection to the TSDB
func (c *TSDBClient) Close() error {
	return c.conn.Close()
}

// WriteData writes a single data point to the TSDB
func (c *TSDBClient) WriteData(key string, timestamp int64, value float64) error {
	_, err := fmt.Fprintf(c.conn, "%s,%d,%.2f\n", key, timestamp, value)
	return err
}

// ReadData reads data from the TSDB for a given key, time range, and downsampling
func (c *TSDBClient) ReadData(key string, startTime, endTime int64, downsampling int) ([]string, error) {
	_, err := fmt.Fprintf(c.conn, "%s,%d,%d,%d\n", key, startTime, endTime, downsampling)
	if err != nil {
		return nil, err
	}

	reader := bufio.NewReader(c.conn)
	response, err := reader.ReadString('\n')
	if err != nil {
		return nil, err
	}

	return strings.Split(strings.TrimSpace(response), "|"), nil
}

// Subscribe subscribes to updates for a given key
func (c *TSDBClient) Subscribe(key string) error {
	_, err := fmt.Fprintf(c.conn, "subscribe,%s\n", key)
	return err
}

// Unsubscribe unsubscribes from updates for a given key
func (c *TSDBClient) Unsubscribe(key string) error {
	_, err := fmt.Fprintf(c.conn, "unsubscribe,%s\n", key)
	return err
}

// RecordMeasurement records a single measurement for a given sensor
func (c *TSDBClient) RecordMeasurement(sensorID string, value float64) error {
	timestamp := time.Now().Unix()
	return c.WriteData(sensorID, timestamp, value)
}

// GetLatestMeasurement retrieves the most recent measurement for a given sensor
func (c *TSDBClient) GetLatestMeasurement(sensorID string) (float64, time.Time, error) {
	endTime := time.Now().Unix()
	startTime := endTime - 3600 // Look back 1 hour to find the latest measurement

	data, err := c.ReadData(sensorID, startTime, endTime, 0)
	if err != nil {
		return 0, time.Time{}, err
	}

	if len(data) == 0 {
		return 0, time.Time{}, fmt.Errorf("no data found for sensor %s", sensorID)
	}

	// Parse the most recent measurement
	parts := strings.Split(data[len(data)-1], ",")
	if len(parts) != 3 {
		return 0, time.Time{}, fmt.Errorf("invalid data format")
	}

	timestamp, err := strconv.ParseInt(parts[1], 10, 64)
	if err != nil {
		return 0, time.Time{}, err
	}

	value, err := strconv.ParseFloat(parts[2], 64)
	if err != nil {
		return 0, time.Time{}, err
	}

	return value, time.Unix(timestamp, 0), nil
}

// GetAverageMeasurement calculates the average measurement over a specified time period
func (c *TSDBClient) GetAverageMeasurement(sensorID string, duration time.Duration) (float64, error) {
	endTime := time.Now().Unix()
	startTime := endTime - int64(duration.Seconds())

	data, err := c.ReadData(sensorID, startTime, endTime, 0)
	if err != nil {
		return 0, err
	}

	if len(data) == 0 {
		return 0, fmt.Errorf("no data found for sensor %s in the specified time range", sensorID)
	}

	var sum float64
	var count int

	for _, measurement := range data {
		parts := strings.Split(measurement, ",")
		if len(parts) != 3 {
			continue
		}

		value, err := strconv.ParseFloat(parts[2], 64)
		if err != nil {
			continue
		}

		sum += value
		count++
	}

	if count == 0 {
		return 0, fmt.Errorf("no valid measurements found")
	}

	return sum / float64(count), nil
}

// GetMeasurementHistory retrieves the measurement history for a given sensor and time range
func (c *TSDBClient) GetMeasurementHistory(sensorID string, startTime, endTime time.Time, interval time.Duration) ([]struct {
	Timestamp time.Time
	Value     float64
}, error) {
	downsamplingSeconds := int(interval.Seconds())
	if downsamplingSeconds < 1 {
		downsamplingSeconds = 1
	}

	data, err := c.ReadData(sensorID, startTime.Unix(), endTime.Unix(), downsamplingSeconds)
	if err != nil {
		return nil, err
	}

	var history []struct {
		Timestamp time.Time
		Value     float64
	}

	for _, measurement := range data {
		parts := strings.Split(measurement, ",")
		if len(parts) != 3 {
			continue
		}

		timestamp, err := strconv.ParseInt(parts[1], 10, 64)
		if err != nil {
			continue
		}

		value, err := strconv.ParseFloat(parts[2], 64)
		if err != nil {
			continue
		}

		history = append(history, struct {
			Timestamp time.Time
			Value     float64
		}{
			Timestamp: time.Unix(timestamp, 0),
			Value:     value,
		})
	}

	return history, nil
}

// Example usage
func example() {
	client, err := NewTSDBClient("localhost:8080")
	if err != nil {
		panic(err)
	}
	defer client.Close()

	// Record a measurement
	err = client.RecordMeasurement("sensor1", 25.5)
	if err != nil {
		panic(err)
	}

	// Get the latest measurement
	value, timestamp, err := client.GetLatestMeasurement("sensor1")
	if err != nil {
		panic(err)
	}
	fmt.Printf("Latest measurement for sensor1: %.2f at %s\n", value, timestamp)

	// Get the average measurement over the last hour
	avgValue, err := client.GetAverageMeasurement("sensor1", time.Hour)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Average measurement for sensor1 over the last hour: %.2f\n", avgValue)

	// Get measurement history for the last 24 hours, with 1-hour intervals
	endTime := time.Now()
	startTime := endTime.Add(-24 * time.Hour)
	history, err := client.GetMeasurementHistory("sensor1", startTime, endTime, time.Hour)
	if err != nil {
		panic(err)
	}
	fmt.Println("Measurement history for sensor1:")
	for _, entry := range history {
		fmt.Printf("  %s: %.2f\n", entry.Timestamp, entry.Value)
	}
}

File: gtsdbDriver.js
const net = require('net');

class TSDBClient {
  constructor(host, port) {
    this.host = host;
    this.port = port;
    this.socket = new net.Socket();
    this.connected = false;
  }

  connect() {
    return new Promise((resolve, reject) => {
      this.socket.connect(this.port, this.host, () => {
        this.connected = true;
        resolve();
      });
      this.socket.on('error', (err) => {
        reject(err);
      });
    });
  }

  close() {
    return new Promise((resolve) => {
      this.socket.end(() => {
        this.connected = false;
        resolve();
      });
    });
  }

  writeData(key, timestamp, value) {
    return new Promise((resolve, reject) => {
      this.socket.write(`${key},${timestamp},${value}\n`, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  readData(key, startTime, endTime, downsampling) {
    return new Promise((resolve, reject) => {
      this.socket.write(`${key},${startTime},${endTime},${downsampling}\n`, (err) => {
        if (err) reject(err);
      });

      this.socket.once('data', (data) => {
        resolve(data.toString().trim().split('|'));
      });
    });
  }

  async recordMeasurement(sensorID, value) {
    const timestamp = Math.floor(Date.now() / 1000);
    await this.writeData(sensorID, timestamp, value);
  }

  async getLatestMeasurement(sensorID) {
    const endTime = Math.floor(Date.now() / 1000);
    const startTime = endTime - 3600; // Look back 1 hour

    const data = await this.readData(sensorID, startTime, endTime, 0);
    if (data.length === 0) {
      throw new Error(`No data found for sensor ${sensorID}`);
    }

    const latestMeasurement = data[data.length - 1].split(',');
    return {
      value: parseFloat(latestMeasurement[2]),
      timestamp: new Date(parseInt(latestMeasurement[1]) * 1000)
    };
  }

  async getAverageMeasurement(sensorID, durationMs) {
    const endTime = Math.floor(Date.now() / 1000);
    const startTime = endTime - Math.floor(durationMs / 1000);

    const data = await this.readData(sensorID, startTime, endTime, 0);
    if (data.length === 0) {
      throw new Error(`No data found for sensor ${sensorID} in the specified time range`);
    }

    let sum = 0;
    let count = 0;

    for (const measurement of data) {
      const parts = measurement.split(',');
      if (parts.length === 3) {
        sum += parseFloat(parts[2]);
        count++;
      }
    }

    if (count === 0) {
      throw new Error('No valid measurements found');
    }

    return sum / count;
  }

  async getMeasurementHistory(sensorID, startTime, endTime, intervalMs) {
    const downsamplingSeconds = Math.max(1, Math.floor(intervalMs / 1000));
    const startTimestamp = Math.floor(startTime.getTime() / 1000);
    const endTimestamp = Math.floor(endTime.getTime() / 1000);

    const data = await this.readData(sensorID, startTimestamp, endTimestamp, downsamplingSeconds);

    return data.map(measurement => {
      const parts = measurement.split(',');
      return {
        timestamp: new Date(parseInt(parts[1]) * 1000),
        value: parseFloat(parts[2])
      };
    });
  }

  subscribe(key) {
    return new Promise((resolve, reject) => {
      this.socket.write(`subscribe,${key}\n`, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  unsubscribe(key) {
    return new Promise((resolve, reject) => {
      this.socket.write(`unsubscribe,${key}\n`, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  onSubscriptionData(callback) {
    this.socket.on('data', (data) => {
      const messages = data.toString().trim().split('\n');
      messages.forEach(message => {
        const [key, timestamp, value] = message.split(',');
        callback({
          key,
          timestamp: new Date(parseInt(timestamp) * 1000),
          value: parseFloat(value)
        });
      });
    });
  }
}

// Example usage
async function example() {
  const client = new TSDBClient('localhost', 8080);

  try {
    await client.connect();

    // Subscribe to updates for sensor1
    await client.subscribe('sensor1');
    console.log('Subscribed to sensor1');

    // Set up a callback for subscription data
    client.onSubscriptionData((data) => {
      console.log(`Received update for ${data.key}: ${data.value} at ${data.timestamp}`);
    });

    // Wait for some time to receive updates
    await new Promise(resolve => setTimeout(resolve, 30000)); // Wait for 30 seconds

    // Unsubscribe from updates for sensor1
    await client.unsubscribe('sensor1');
    console.log('Unsubscribed from sensor1');

    // Record a measurement
    await client.recordMeasurement('sensor1', 25.5);
    console.log('Measurement recorded');

    // Get the latest measurement
    const latestMeasurement = await client.getLatestMeasurement('sensor1');
    console.log(`Latest measurement for sensor1: ${latestMeasurement.value} at ${latestMeasurement.timestamp}`);

    // Get the average measurement over the last hour
    const avgValue = await client.getAverageMeasurement('sensor1', 3600000); // 1 hour in milliseconds
    console.log(`Average measurement for sensor1 over the last hour: ${avgValue}`);

    // Get measurement history for the last 24 hours, with 1-hour intervals
    const endTime = new Date();
    const startTime = new Date(endTime.getTime() - 24 * 60 * 60 * 1000);
    const history = await client.getMeasurementHistory('sensor1', startTime, endTime, 3600000); // 1 hour in milliseconds
    console.log('Measurement history for sensor1:');
    history.forEach(entry => {
      console.log(`  ${entry.timestamp}: ${entry.value}`);
    });

  } catch (error) {
    console.error('Error:', error);
  } finally {
    await client.close();
  }
}
